x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java

```diff
@@ -929,12 +929,49 @@ void populate(
                         listener.onResponse(false);
                         return;
                     }
-                    try (var gapsListener = new RefCountingListener(listener.map(unused -> true))) {
-                        assert writer.sharedInputStreamFactory(gaps) == null;
-                        for (SparseFileTracker.Gap gap : gaps) {
-                            executor.execute(
-                                fillGapRunnable(gap, writer, null, ActionListener.releaseAfter(gapsListener.acquire(), refs.acquire()))
-                            );
+                    final SourceInputStreamFactory streamFactory = writer.sharedInputStreamFactory(gaps);
+                    logger.trace(
+                        () -> Strings.format(
+                            "fill gaps %s %s shared input stream factory",
+                            gaps,
+                            (streamFactory == null ? "without" : "with"),
+                            (streamFactory == null ? "" : " " + streamFactory)
+                        )
+                    );
+                    if (streamFactory == null) {
+                        try (var parallelGapsListener = new RefCountingListener(listener.map(unused -> true))) {
+                            for (SparseFileTracker.Gap gap : gaps) {
+                                executor.execute(
+                                    fillGapRunnable(
+                                        gap,
+                                        writer,
+                                        null,
+                                        ActionListener.releaseAfter(parallelGapsListener.acquire(), refs.acquire())
+                                    )
+                                );
+                            }
+                        }
+                    } else {
+                        try (
+                            var sequentialGapListener = new RefCountingListener(
```

Comment:
Nit
```suggestion
                            var sequentialGapsListener = new RefCountingListener(
```
to match `parallelGapsListener`